File: src/main/java/net/unaussprechlich/managedgui/lib/util/ColorRGBA.java
/*
 * ***************************************************************************
 *
 *         Copyright © 2016 unaussprechlich - ALL RIGHTS RESERVED
 *
 * ***************************************************************************
 */
package net.unaussprechlich.managedgui.lib.util;


public class ColorRGBA {

    private final int RED;
    private final int GREEN;
    private final int BLUE;
    private final int ALPHA;

    private final float REDf;
    private final float GREENf;
    private final float BLUEf;
    private final float ALPHAf;

    public ColorRGBA(int RED, int GREEN, int BLUE, int ALPHA) {
        this.RED   = RED;
        this.GREEN = GREEN;
        this.BLUE  = BLUE;
        this.ALPHA = ALPHA;

        this.REDf   = ((float)RED)   / 255;
        this.GREENf = ((float)GREEN) / 255;
        this.BLUEf  = ((float)BLUE)  / 255;
        this.ALPHAf = ((float)ALPHA) / 255;
    }

    public ColorRGBA(int RED, int GREEN, int BLUE) {
        int ALPHA = 255;
        this.RED   = RED;
        this.GREEN = GREEN;
        this.BLUE  = BLUE;
        this.ALPHA = ALPHA;

        this.REDf   = ((float)RED)   / 255;
        this.GREENf = ((float)GREEN) / 255;
        this.BLUEf  = ((float)BLUE)  / 255;
        this.ALPHAf = ((float)ALPHA) / 255;
    }

    public ColorRGBA(float RED, float GREEN, float BLUE, float ALPHA) {
        this.RED   = Math.round(RED   * 255f);
        this.GREEN = Math.round(GREEN * 255f);
        this.BLUE  = Math.round(BLUE  * 255f);
        this.ALPHA = Math.round(ALPHA * 255f);

        this.REDf   = RED;
        this.GREENf = GREEN;
        this.BLUEf  = BLUE;
        this.ALPHAf = ALPHA;
    }

    public ColorRGBA(float RED, float GREEN, float BLUE) {
        float ALPHA = 1f;
        this.RED   = Math.round(RED   * 255f);
        this.GREEN = Math.round(GREEN * 255f);
        this.BLUE  = Math.round(BLUE  * 255f);
        this.ALPHA = Math.round(ALPHA * 255f);

        this.REDf   = RED;
        this.GREENf = GREEN;
        this.BLUEf  = BLUE;
        this.ALPHAf = ALPHA;
    }

    public float getALPHAf() {return ALPHAf;}
    public float getBLUEf()  {return BLUEf;}
    public float getGREENf() {return GREENf;}
    public float getREDf()   {return REDf;}

    public int getALPHA()  {return ALPHA;}
    public int getBLUE()   {return BLUE;}
    public int getGREEN()  {return GREEN;}
    public int getRED()    {return RED;}

    public static int interpolateColor(int rgba1, int rgba2, float percent) {
        int r1 = rgba1 & 0xFF, g1 = rgba1 >> 8 & 0xFF, b1 = rgba1 >> 16 & 0xFF, a1 = rgba1 >> 24 & 0xFF;
        int r2 = rgba2 & 0xFF, g2 = rgba2 >> 8 & 0xFF, b2 = rgba2 >> 16 & 0xFF, a2 = rgba2 >> 24 & 0xFF;

        int r = (int) (r1 < r2 ? r1 + (r2 - r1) * percent : r2 + (r1 - r2) * percent);
        int g = (int) (g1 < g2 ? g1 + (g2 - g1) * percent : g2 + (g1 - g2) * percent);
        int b = (int) (b1 < b2 ? b1 + (b2 - b1) * percent : b2 + (b1 - b2) * percent);
        int a = (int) (a1 < a2 ? a1 + (a2 - a1) * percent : a2 + (a1 - a2) * percent);

        return r | g << 8 | b << 16 | a << 24;
    }

    public static int toRGBA(ColorRGBA c) {
        return c.getRED() | c.getGREEN() << 8 | c.getBLUE() << 16 | c.getALPHA() << 24;
    }

    public static ColorRGBA toColor(int rgba) {
        int r = rgba & 0xFF, g = rgba >> 8 & 0xFF, b = rgba >> 16 & 0xFF, a = rgba >> 24 & 0xFF;
        return new ColorRGBA(r, g, b, a);
    }

    public static ColorRGBA getColorFromFormatting(char c){
        ColorRGBA color = RGBA.WHITE_MC.get();
        switch (c){
            case '0': color = RGBA.BLACK_MC.get(); break;
            case '1': color = RGBA.BLUE_DARK_MC.get(); break;
            case '2': color = RGBA.GREEN_DARK_MC.get(); break;
            case '3': color = RGBA.AQUA_DARK_MC.get(); break;
            case '4': color = RGBA.RED_DARK_MC.get(); break;
            case '5': color = RGBA.PURPLE_DARK_MC.get(); break;
            case '6': color = RGBA.GOLD_MC.get(); break;
            case '7': color = RGBA.GRAY_MC.get(); break;
            case '8': color = RGBA.GRAY_DARK_MC.get(); break;
            case '9': color = RGBA.BLUE_MC.get(); break;
            case 'a': color = RGBA.GREEN_MC.get(); break;
            case 'b': color = RGBA.AQUA_MC.get(); break;
            case 'c': color = RGBA.RED_MC.get(); break;
            case 'd': color = RGBA.PURPLE_LIGHT_MC.get(); break;
            case 'e': color = RGBA.YELLOW_MC.get(); break;
            case 'f': color = RGBA.WHITE_MC.get(); break;
        }
        return color;
    }
}

File: src/main/java/net/unaussprechlich/managedgui/lib/util/ColorRGBA.kt
/*
 * ***************************************************************************
 *
 *         Copyright © 2016 unaussprechlich - ALL RIGHTS RESERVED
 *
 * ***************************************************************************
 */
package net.unaussprechlich.managedgui.lib.util


class ColorRGBA {

    val red: Int
    val green: Int
    val blue: Int
    val alpha: Int

    val redf: Float
    val greenf: Float
    val bluef: Float
    val alphaf: Float

    constructor(RED: Int, GREEN: Int, BLUE: Int, ALPHA: Int) {
        this.red = RED
        this.green = GREEN
        this.blue = BLUE
        this.alpha = ALPHA

        this.redf = RED.toFloat() / 255
        this.greenf = GREEN.toFloat() / 255
        this.bluef = BLUE.toFloat() / 255
        this.alphaf = ALPHA.toFloat() / 255
    }

    constructor(RED: Int, GREEN: Int, BLUE: Int) {
        val ALPHA = 255
        this.red = RED
        this.green = GREEN
        this.blue = BLUE
        this.alpha = ALPHA

        this.redf = RED.toFloat() / 255
        this.greenf = GREEN.toFloat() / 255
        this.bluef = BLUE.toFloat() / 255
        this.alphaf = ALPHA.toFloat() / 255
    }

    constructor(RED: Float, GREEN: Float, BLUE: Float, ALPHA: Float) {
        this.red = Math.round(RED * 255f)
        this.green = Math.round(GREEN * 255f)
        this.blue = Math.round(BLUE * 255f)
        this.alpha = Math.round(ALPHA * 255f)

        this.redf = RED
        this.greenf = GREEN
        this.bluef = BLUE
        this.alphaf = ALPHA
    }

    constructor(RED: Float, GREEN: Float, BLUE: Float) {
        val ALPHA = 1f
        this.red = Math.round(RED * 255f)
        this.green = Math.round(GREEN * 255f)
        this.blue = Math.round(BLUE * 255f)
        this.alpha = Math.round(ALPHA * 255f)

        this.redf = RED
        this.greenf = GREEN
        this.bluef = BLUE
        this.alphaf = ALPHA
    }

    companion object {

        fun interpolateColor(rgba1: Int, rgba2: Int, percent: Float): Int {
            val r1 = rgba1 and 0xFF
            val g1 = rgba1 shr 8 and 0xFF
            val b1 = rgba1 shr 16 and 0xFF
            val a1 = rgba1 shr 24 and 0xFF
            val r2 = rgba2 and 0xFF
            val g2 = rgba2 shr 8 and 0xFF
            val b2 = rgba2 shr 16 and 0xFF
            val a2 = rgba2 shr 24 and 0xFF

            val r = (if (r1 < r2) r1 + (r2 - r1) * percent else r2 + (r1 - r2) * percent).toInt()
            val g = (if (g1 < g2) g1 + (g2 - g1) * percent else g2 + (g1 - g2) * percent).toInt()
            val b = (if (b1 < b2) b1 + (b2 - b1) * percent else b2 + (b1 - b2) * percent).toInt()
            val a = (if (a1 < a2) a1 + (a2 - a1) * percent else a2 + (a1 - a2) * percent).toInt()

            return r or (g shl 8) or (b shl 16) or (a shl 24)
        }

        fun toRGBA(c: ColorRGBA): Int {
            return c.red or (c.green shl 8) or (c.blue shl 16) or (c.alpha shl 24)
        }

        fun toColor(rgba: Int): ColorRGBA {
            val r = rgba and 0xFF
            val g = rgba shr 8 and 0xFF
            val b = rgba shr 16 and 0xFF
            val a = rgba shr 24 and 0xFF
            return ColorRGBA(r, g, b, a)
        }

        fun getColorFromFormatting(c: Char): ColorRGBA {
            var color = RGBA.WHITE_MC.get()
            when (c) {
                '0' -> color = RGBA.BLACK_MC.get()
                '1' -> color = RGBA.BLUE_DARK_MC.get()
                '2' -> color = RGBA.GREEN_DARK_MC.get()
                '3' -> color = RGBA.AQUA_DARK_MC.get()
                '4' -> color = RGBA.RED_DARK_MC.get()
                '5' -> color = RGBA.PURPLE_DARK_MC.get()
                '6' -> color = RGBA.GOLD_MC.get()
                '7' -> color = RGBA.GRAY_MC.get()
                '8' -> color = RGBA.GRAY_DARK_MC.get()
                '9' -> color = RGBA.BLUE_MC.get()
                'a' -> color = RGBA.GREEN_MC.get()
                'b' -> color = RGBA.AQUA_MC.get()
                'c' -> color = RGBA.RED_MC.get()
                'd' -> color = RGBA.PURPLE_LIGHT_MC.get()
                'e' -> color = RGBA.YELLOW_MC.get()
                'f' -> color = RGBA.WHITE_MC.get()
            }
            return color
        }
    }
}
